File: scripts/Input.cs
/*
* Copyright 2021 Rochus Keller <mailto:me@rochus-keller.ch>
*
* This file is part of the Oberon+ parser/compiler library.
*
* The following is the license that applies to this copy of the
* library. For a license to use the library under conditions
* other than those described here, please email to me@rochus-keller.ch.
*
* GNU General Public License Usage
* This file may be used under the terms of the GNU General Public
* License (GPL) versions 2.0 or 3.0 as published by the Free Software
* Foundation and appearing in the file LICENSE.GPL included in
* the packaging of this file. Please review the following information
* to ensure GNU General Public Licensing requirements will be met:
* http://www.fsf.org/licensing/licenses/info/GPLv2.html and
* http://www.gnu.org/copyleft/gpl.html.
*/

public class Input
{
	//CONST TimeUnit = 1000000;
	
	//PROCEDURE Available (): INTEGER;
	private static int cur = -1;
	public static int Available()
	{
		if( cur >= 0 )
			return 1;
		cur = System.Console.Read();
		if( cur >= 0 )
			return 1;
		else
			return 0;
	}
	
	//PROCEDURE Read (VAR ch: CHAR);
	public static void Read(ref char ch)
	{
		if( cur >= 0 )
		{
			ch = (char)cur;
			cur = -1;
		}else
			ch = '\0';
	}
	
	//PROCEDURE Mouse (VAR keys: SET; VAR x, y: INTEGER);
	public static void Mouse( ref uint keys, ref int x, ref int y)
	{
		// TODO
	}
	
	//PROCEDURE SetMouseLimits (w, h: INTEGER);
	public static void SetMouseLimits(int w, int h)
	{
		// NOP
	}
	
	static private System.DateTime start = System.DateTime.Now;
	
	//PROCEDURE Time (): INTEGER;
	public static int Time()
	{
		// A single tick represents one hundred nanoseconds or one ten-millionth of a second. There are 10,000 ticks in a millisecond.
		return (int)( ( System.DateTime.Now.Ticks - start.Ticks ) / 10 ); // microseconds
	}
	
}

File: scripts/Input.c
/*
* Copyright 2021 Rochus Keller <mailto:me@rochus-keller.ch>
*
* This file is part of the Oberon+ parser/compiler library.
*
* The following is the license that applies to this copy of the
* library. For a license to use the library under conditions
* other than those described here, please email to me@rochus-keller.ch.
*
* GNU General Public License Usage
* This file may be used under the terms of the GNU General Public
* License (GPL) versions 2.0 or 3.0 as published by the Free Software
* Foundation and appearing in the file LICENSE.GPL included in
* the packaging of this file. Please review the following information
* to ensure GNU General Public Licensing requirements will be met:
* http://www.fsf.org/licensing/licenses/info/GPLv2.html and
* http://www.gnu.org/copyleft/gpl.html.
*/

#include "Input.h"
#include <time.h>

#if defined(_WIN32) && !defined(__GNUC__)
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

// Source: https://stackoverflow.com/questions/10905892/equivalent-of-gettimeday-for-windows/26085827

// MSVC defines this in winsock2.h!?
typedef struct timeval {
    long tv_sec;
    long tv_usec;
} timeval;

int gettimeofday(struct timeval * tp, struct timezone * tzp)
{
    // Note: some broken versions only have 8 trailing zero's, the correct epoch has 9 trailing zero's
    // This magic number is the number of 100 nanosecond intervals since January 1, 1601 (UTC)
    // until 00:00:00 January 1, 1970 
    static const uint64_t EPOCH = ((uint64_t) 116444736000000000ULL);

    SYSTEMTIME  system_time;
    FILETIME    file_time;
    uint64_t    time;

    GetSystemTime( &system_time );
    SystemTimeToFileTime( &system_time, &file_time );
    time =  ((uint64_t)file_time.dwLowDateTime )      ;
    time += ((uint64_t)file_time.dwHighDateTime) << 32;

    tp->tv_sec  = (long) ((time - EPOCH) / 10000000L);
    tp->tv_usec = (long) (system_time.wMilliseconds * 1000);
    return 0;
}
#else
#include <sys/time.h>
#endif

static struct timeval start;

int32_t Input$Time()
{
#ifdef _USE_CPU_TIME_
	clock_t now = clock();
	if( now < 0 )
		return 0;
	return now * 1000000 / CLOCKS_PER_SEC;
#else
    static struct timeval now;
    gettimeofday(&now, 0);
    const long seconds = now.tv_sec - start.tv_sec;
    const long microseconds = now.tv_usec - start.tv_usec;
    return seconds*1000000 + microseconds;
#endif
}

void Input$init$()
{
    gettimeofday(&start, 0);
}

OBX$Cmd Input$cmd$(const char* name)
{
	if( name == 0 ) return Input$init$;
	return 0;
}
